core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -311,16 +313,52 @@ class BlockManagerMasterEndpoint(
   }
 
   private def removeShuffle(shuffleId: Int): Future[Seq[Boolean]] = {
-    // Nothing to do in the BlockManagerMasterEndpoint data structures
+    // Find all shuffle blocks on executors that are no longer running
+    val blocksToDeleteByShuffleService =
+      new mutable.HashMap[BlockManagerId, mutable.HashSet[BlockId]]
+    if (externalBlockStoreClient.isDefined) {
+      mapOutputTracker.shuffleStatuses.get(shuffleId).foreach { shuffleStatus =>
+        shuffleStatus.mapStatuses.foreach { mapStatus =>
+          // Port should always be external shuffle port if external shuffle is enabled
+          val isShufflePort = mapStatus.location.port == externalShuffleServicePort
+          val executorDeallocated =
+            !blockManagerIdByExecutor.contains(mapStatus.location.executorId)
+          if (isShufflePort && executorDeallocated) {
+            val blocks = blocksToDeleteByShuffleService.getOrElseUpdate(mapStatus.location,
+              new mutable.HashSet[BlockId])
```

Comment:
get/update only if `blocksToDel` is non-empty (if we have possibility of other `ShuffleBlockResolver`)

Reply:
good point, updated